### PR TITLE
Major refactoring and gradient fix 

### PR DESCRIPTION
**Pull request by marvinfriede**
- adapt Fortran code structure
- gradient is now correct and translationally invariant
- test gradient against numerical gradient

### CODE REVIEW

**Comment by marvinfriede** [comment]
Note that this new structure makes the code somewhat slower.
I suspect that this comes from calculating the reference C6 coefficients separately for the two- and the three-body term. Before, they were calculated "on-the-fly" within the energy/gradient functions. The "on-the-fly" version, however, was very difficult to debug as it is very different from the Fortran reference.

File: include/dftd_dispersion.h [hunk 1]
```diff
@@ -38,53 +38,135 @@ class dparam {
...
- * @param charge Charge of the molecule.
  * @param cutoff Real-space cutoffs for CN and dispersion.
  * @param energy Dispersion energy.
  * @param GRAD Dispersion gradient.
  * @return Exit status.
  */
-extern
-int get_dispersion(const TMolecule &mol, const dparam &par, const int &charge,
-                   TCutoff cutoff, double &energy, double *GRAD);
+extern int DFTVDW_D4(
```
Comment (awvwgk) [inline include/dftd_dispersion.h:52]: This naming doesn't match the rest of the functions. I suggest to stick with snake_case naming conventions.

File: include/dftd_geometry.h [hunk 2]
```diff
@@ -44,4 +44,13 @@ class TMolecule {
...
 
-}  // namespace dftd
+class TMolInfo {
+  public:
+    int Charge;
+
+    TMolInfo(int c = 0) {
+      Charge = c;
+    }
+};
```
Comment (awvwgk) [inline include/dftd_geometry.h:54]: Can this be part of the `TMolecule` class?

File: include/dftd_cblas.h [hunk 3]
```diff
@@ -24,24 +24,24 @@ extern "C" {
 #include "lapacke.h"
 }
 
-namespace dftd {
+namespace dftd4 {
```
[the remark this answers is not in the document]
Comment (marvinfriede) [reply include/dftd_cblas.h:27]: Not necessarily, that was just an adaptation to the ORCA code.

File: include/dftd_geometry.h (hunk 2 above)
Comment (marvinfriede) [reply include/dftd_geometry.h:54]: This was again an adaptation to the ORCA code which handles geometry and charge in different classes. I did not want to re-write everything when implementing in ORCA, so I created this dummy `TMolInfo` class. I am not quite sure what the best course of action would be here.

File: include/dftd_dispersion.h (hunk 1 above)
Comment (marvinfriede) [reply include/dftd_dispersion.h]: Agreed, renamed to `get_dispersion`.

File: include/dftd_geometry.h (hunk 2 above)
Comment (marvinfriede) [reply include/dftd_geometry.h]: The charge is now passed as `int` again. The ORCA specifics will be handled in a dedicated wrapper only within ORCA.

**Review by awvwgk — APPROVED** [review]

**Pull request merged**